### ISSUE TITLE
Add Basic Random Input Generation to CW Testing Framework

### ISSUE THREAD

**Issue by suic86** [codewars/codewars.com#781]
## Feature
Add basic functions for random input generation.

## Rationale
On the one hand, random tests have become _de facto_ standard for any new kata and missing random tests are often reported as an issues. On the other hand, CW doesn't offer any tooling for random input creation. Random input creation can be challenging for newbies (e. g. in Python it's still relatively simple but the documentation of the [random](https://docs.python.org/3/library/random.html?highlight=random.choice#random.choice) module isn't the easiest thing to read) and annoying for experienced users (as it's about writing or copying the same random input generation functions over and over again). Adding a set of basic random generation function would improve both (e.g. myjinxin2015 has a nice miniframework for JS and in Python it would be simple too :))

## Examples (in Python)
```python
def random_string(length=10, character=ascii_lowercase):
      return "".join(random.choice(chars) for _ in range(length))


def random_number(lower_bound=0, upper_bound=10):
     return random.randint(lower_bound, upper_bound)
```

┆Issue is synchronized with this [Clickup ](https://app.clickup.com/t/2me1fbd) by [Unito](https://www.unito.io)

**Comment by jhoffner** [comment]
FYI there are some randomization features for JS. 

`Test.randomize(array)` - great if you just want to randomized the order of tests
`Test.randomNumber()`
`Test.randomToken()` - generate a string
`Test.sample()` - useful for predefining an array and then randomly selecting a value from it.

Also these features would be added into the [cli project](https://github.com/Codewars/codewars-runner-cli). I doubt I'm ever going to get to adding them in myself but its open for contributions

**Comment by kazk** [comment]
@jhoffner, do you have contributing guidelines for [codewars-runner-cli](https://github.com/Codewars/codewars-runner-cli)? I'm thinking of contributing to the test framework, but worried because I have no experience with Docker. How much Docker knowledge do I need?

**Comment by jhoffner** [comment]
Little docker knowledge is needed for the most part. The readme has a bunch of info on contributing. 

What were you looking to add?

**Comment by kazk** [comment]
I read the readme, but wasn't sure if all that setup was necessary for fixing existing code (not adding new language support).
I also didn't know where to add tests. I didn't see tests for functions you listed above in [/test/runners/javascript_spec.js](https://github.com/Codewars/codewars-runner-cli/blob/master/test/runners/javascript_spec.js). If local test is not a requirement, I can start contributing sooner.

> What were you looking to add?

I'm thinking of contributing to [/frameworks/javascript/cw-2.js](https://github.com/Codewars/codewars-runner-cli/blob/master/frameworks/javascript/cw-2.js) as a start and anywhere help is wanted (including other languages). For example, `Test.randomNumber()` is not uniform:

```javascript
// in kumite Node v6.6.0
const R = Array.from({length: 101}, _ => 0); 
const F = Array.from({length: 101}, _ => 0); 
for (let i = 0; i < 1e7; ++i) {
  ++R[Math.round(Math.random()*100)]; // Test.randomNumber()
  ++F[Math.floor(Math.random()*101)];
}
console.log(R.map((x, i) => [i, 100*x/1e7]));
//=> [0, 0.49823],[1, 0.99786],[2, 1.00047],...,[98, 1.00617],[99, 0.99226],[100, 0.49933]
console.log(F.map((x, i) => [i, 100*x/1e7]));
//=> [0, 0.98429],[1, 0.98672],[2, 0.98790],...,[98, 0.98940],[99, 0.99204],[100, 0.99483]
```

**Comment by jhoffner** [comment]
For the most part you just need to: 
- clone the repo
- install docker
- run `make node`
- edit `cw-2.js` and `javascript_spec.js`
- run `docker-compose run javascript_test`

**Comment by kazk** [comment]
Thanks! Looks much easier than I thought.
One more question. Is there a reason for not using `lodash` for above functions? I saw `lodash` used for `deepEquals` on line 5.

- `Test.randomize(array)` is `_.shuffle(array)`
- `Test.randomNumber()` is `_.random(100)`
- `Test.sample(array)` is `_.sample(array)`

**Comment by jhoffner** [comment]
That part of the framework was written before lodash (or any other 3rd party require) was supported.

**Comment by kazk** [comment]
I see :)

I've been trying to set up Docker and I was able to build the base and node image. I could not get `docker-compose run javascript_test` to find the test file, so I went in with bash and ran tests:
- `docker run --rm -it --entrypoint bash codewars/node-runner`
- `codewarrior@c4e6efe5208e:/runner$ mocha -t 5000 test/runners/javascript_spec`
- `58 passing (29s)`

Is this ok?

**Comment by jhoffner** [comment]
Yeah you can use `docker run` yourself but unless you also map the volumes, you will not pick up any updates that you make. As you continue to work on improvements, I would highly suggest figuring out what is wrong so that you can get docker-compose to work - it will make your dev flow 100x better.

**Comment by kazk** [comment]
I couldn't figure out so I did `make node` each time I changed files 😑 

> unless you also map the volumes

Do you mean with `-v` option? I tried the command from README.

```
codewas-runner-cli$ docker run -it --rm --entrypoint bash \
        -v $(pwd)/lib:/runner/lib \
        -v $(pwd)/examples:/runner/examples \
        -v $(pwd)/frameworks:/runner/frameworks \
        -v $(pwd)/test:/runner/test \
        codewars/node-runner

codewarrior@69f26bb00ca1:/runner$ mocha -t 5000 test/runners/javascript_spec

Warning: Could not find any test files matching pattern: test/runners/javascript_spec
No test files found

codewarrior@69f26bb00ca1:/runner$ ls test # empty
codewarrior@69f26bb00ca1:/runner$ ls lib # empty
codewarrior@69f26bb00ca1:/runner$ ls frameworks # empty
codewarrior@69f26bb00ca1:/runner$ exit

docker -v # Docker version 1.12.5, build 7392c3b
```
I think `docker-compose` showed the same warning.

It worked if I used relative path (`name` value? in Docker doc) for `host-src`:

```
codewas-runner-cli$ docker run -it --rm --entrypoint bash \
        -v lib:/runner/lib \
        -v examples:/runner/examples \
        -v frameworks:/runner/frameworks \
        -v test:/runner/test \
        codewars/node-runner
codewarrior@568695ff452a:/runner$ ls test
# runner.js  runners
codewarrior@568695ff452a:/runner$  mocha -t 5000 test/runners/javascript_spec
#  64 passing (38s)
```

> The `container-dest` must always be an absolute path such as `/src/docs`. The `host-src` can either be an absolute path or a `name` value. If you supply an absolute path for the `host-dir`, Docker bind-mounts to the path you specify. If you supply a `name`, Docker creates a named volume by that `name`.
> <https://docs.docker.com/engine/reference/run/#/volume-shared-filesystems>

**Comment by kazk** [comment]
I got `docker-compose run javascript_test` working by changing `docker-compose.yml`, but I don't know what I'm doing :neutral_face:

```yaml
# changed
services:
  # ...
  javascript_test:
    image: codewars/node-runner
    volumes:
      - lib:/runner/lib
      - examples:/runner/examples
      - frameworks:/runner/frameworks
      - test:/runner/test
    entrypoint: 'mocha -t 5000 test/runners/javascript_spec.js'

# added
volumes:
  lib: #
  examples: #
  frameworks: #
  test: #
```

**Comment by kazk** [comment]
I still don't know what I'm doing, but I'm guessing my problem is running old OS X and not having `$HOME` under `/Users`, so Docker can't find the path. I didn't realize this multiple layering at first.

`make node` is the only way I've found so far to run the tests I added. I'm going to stop investigating and consider upgrading/cleaning my system which I've been procrastinating for few years 😓 

I ended up learning Docker more than I intended, but it's really cool and I'm glad I tried 😄 
Thanks again for your help.
